### ISSUE TITLE
Understanding relationships between different variables and extracting the factors that cause these relationships

### ISSUE THREAD

**Issue by amjass12** [rstudio/keras#741]
Hi all, 

I am having a bit of an issue with a trained model in that I cannot figure out how to extract information about patterns forming as a result of relationships in the data. Here is what I mean, 

I have built a simple MLP to extract patterns in some genomic time course data I have. I have 1-hot encoded classes such as age, organs, gender etc... I have 5 forms of metadata that have been been 1-hot encoded:
`dataonehot <- data.frame(to_categorical(data$Age, 7), to_categorical(data$Organ, 13), to_categorical(data$Sex, 4),  **etc**, data [,1:5000]` data 1:5000 being the list of genes:

Data is split:

I then specify the one-hot encoded data: 

`trainingtarget1 <- Normcountsonehot[ind==1, c(1:7,8:20,21:24, 25:30,31:33,34:36)] `

and count data: 
`training <- Normcountsonehot[ind==1, 37:5098] # split count data for training [non-one hot encoded data]`

build the model and run: 

model <- keras_model_sequential()
model %>%
   layer_dense(units=150, activation = 'relu', input_shape = 5062) %>%  
   layer_dense(units = 64, activation = "relu") %>%
   layer_dense(units=ncol(trainingtarget1), activation = "sigmoid")

model1 %>% keras::compile(loss='binary_crossentropy',
                          optimizer='adam',
                          metrics='accuracy')

## Run model ##################

early_stop <- callback_early_stopping(monitor = "val_loss", patience = 20)
print_dot_callback <- callback_lambda(
   on_epoch_end = function(epoch, logs) {
      if (epoch %% 80 == 0) cat("\n")
      cat(".")
   }
)    

the model performs well with high-ish accuracy ~89%. and predict classes for the individual one-hot data tell me how well the model performs in identifying those classes:

However what I would like to understand now is the following: what genes cause the machine to light up given relation BETWEEN the one-hot data, and the best example would be something along the lines of:

What genes 'light up' in the machine, because in the young ages, there are 5 genes that are made really highly in 5 organs (out of 10 for example) specifically in females (again, just an example)... This is just an example of a question, however these are the kinds of relationships I want to probe!, effectively, what genes light up when combinations of classes form interesting patterns, i.e, sex specific genes made only in some organs at young ages and not older... Given the richness of this data, it is these patterns I am interesting in understanding beyond just simple classification: 

My question is, how do I extract this information from the model! Any help would really be much appreciated!! Thank you!

**Comment by skeydan** [comment]
Hi, without fully understanding the subject matter, I suspect a neural network might not necessarily be the best tool given the goals of your analysis. I'd suggest asking on the general RStudio ML forum instead:

https://community.rstudio.com/c/ml

**Comment by amjass12** [comment]
Thank you for the reply! I think given that there are complexities in the data (linked to the categorical data) I think that a neural network would be able to extract these patterns as these are hard to extract using other statistical methods... Is there no way of extracting relatonships between classes after models have run?

Many thanks!
